### PR TITLE
Enable hyperlinks that are not defined with w:hyperlink

### PR DESCRIPTION
**Pull request by jdlehman**
Currently hyperlinks defined with `w:hyperlink` work correctly, but other hyperlinks do not.

## Case 1 (working in master)

The following will look up the URI based on the `id` and works.
```xml
         <w:hyperlink r:id="rId5" w:history="1">
            <w:r w:rsidRPr="00E71B0C">
               <w:rPr>
                  <w:rStyle w:val="Hyperlink" />
               </w:rPr>
               <w:t>link</w:t>
            </w:r>
         </w:hyperlink>
```

## Case 2 (fixed by this PR)

The alternate case that this PR seeks to fix is:
```xml
      <w:p w14:paraId="163F3025" w14:textId="77777777" w:rsidR="00E71B0C" w:rsidRPr="00E71B0C" w:rsidRDefault="00E71B0C">
         <w:pPr>
            <w:rPr>
               <w:rStyle w:val="Hyperlink" />
            </w:rPr>
         </w:pPr>
         <w:r>
            <w:fldChar w:fldCharType="begin" />
         </w:r>
         <w:r>
            <w:instrText xml:space="preserve"> HYPERLINK "http://www.google.com/" </w:instrText>
         </w:r>
         <w:r>
            <w:fldChar w:fldCharType="separate" />
         </w:r>
         <w:r w:rsidRPr="00E71B0C">
            <w:rPr>
               <w:rStyle w:val="Hyperlink" />
            </w:rPr>
            <w:t>link to google</w:t>
         </w:r>
         <w:r>
            <w:fldChar w:fldCharType="end" />
         </w:r>
      </w:p>
```

## Solution

Both cases indicate that there is a hyperlink with `w:rStyle`, but the latter stores the URI in `w:instrText`. It seems that `w:instrText` can store a variety of data (`AUTHOR`, `DATE`, `HYPERLINK`, etc). It also appears that `w:fldChar` has an attribute `w:fldCharType` that indicates when that data "begins" and "ends". [Here](http://officeopenxml.com/WPfields.php) is some general information I found on these fields.

Essentially this approach stores `w:instrText` data such that it can be used in the correct context. Right now it is only used for hyperlinks, but we could easily add support for other use case (though the other cases I found like `AUTHOR`, and `DATE`, do not seem as useful to correctly render HTML). Whenever a `w:fldChar` with the attribute `w:fldCharType="end"` is hit, we can clear the `w:instrText` data as this field indicates that we are leaving the current context of this complex field, which prevents context from leaking where it should not be.

I added some tests and the code is fairly minimal, but let me know if you have any thoughts or suggestions for improving this solution.

### CODE REVIEW

File: lib/docx/body-reader.js [hunk 1]
```diff
@@ -6,6 +6,7 @@ var documents = require("../documents");
 var Result = require("../results").Result;
 var warning = require("../results").warning;
 var uriToZipEntryName = require("./uris").uriToZipEntryName;
+var lastInstrText = null;
```
Comment (mwilliamson) [inline lib/docx/body-reader.js]: This introduces global state, which I'd rather avoid.

File: lib/docx/body-reader.js [hunk 2]
```diff
@@ -133,14 +134,27 @@ function BodyReader(options) {
...
                 .map(function(children) {
                     var properties = _.find(children, isRunProperties);
-                    
-                    return new documents.Run(
-                        children.filter(negate(isRunProperties)),
-                        properties
-                    );
+                    var filteredChildren = children.filter(negate(isRunProperties));
+
+                    if (properties && properties.styleName === 'Hyperlink' && lastInstrText) {
```
Comment (mwilliamson) [inline lib/docx/body-reader.js]: Filtering by style name feels brittle to me. It seems like the correct approach to parse the actual text in `w:instrText`.

File: test/docx/body-reader.tests.js [hunk 3]
```diff
@@ -861,6 +873,23 @@ function createLinkedBlip(relationshipId) {
     return new XmlElement("a:blip", {"r:link": relationshipId});
 }
 
+function createHyperlink(link) {
```
Comment (mwilliamson) [inline test/docx/body-reader.tests.js]: Given that there are two ways of creating hyperlinks, the name of this should probably be more explicit so that it describes what sort of hyperlink it's creating.

Comment (mwilliamson) [inline test/docx/body-reader.tests.js]: I'd also limit this function to just creating the XML for the test case, and leaving the reading to the actual test case itself for clarity.

Also, knowing to pluck out the third child feels a bit magic.

File: lib/docx/body-reader.js [hunk 4]
```diff
@@ -491,3 +505,11 @@ function joinElements(first, second) {
 function identity(value) {
     return value;
 }
+
+function clearState() {
+    lastInstrText = null;
+}
+
+function parseHyperlink(instrText) {
```
Comment (mwilliamson) [inline lib/docx/body-reader.js]: Being a bit picky, but `parseHyperlink` feels a bit vague. Specifically, this is actually parsing the `w:instrText` text, whatever the correct term for that is. I think it might be field codes, but the spec should say.

In other words, perhaps something like: `parseHyperlinkFieldCode` (if they are indeed called field codes).

**Comment by mwilliamson** [comment]
Thanks for the pull request. I think this would be good behaviour to merge, and I left a few comments on the implementation.

File: lib/docx/body-reader.js [hunk 5]
```diff
@@ -133,14 +134,27 @@ function BodyReader(options) {
...
                     var properties = _.find(children, isRunProperties);
-                    
-                    return new documents.Run(
-                        children.filter(negate(isRunProperties)),
-                        properties
-                    );
+                    var filteredChildren = children.filter(negate(isRunProperties));
+
+                    if (properties && properties.styleName === 'Hyperlink' && lastInstrText) {
+                        return new documents.Hyperlink(filteredChildren, {href: parseHyperlink(lastInstrText)});
```
Comment (mwilliamson) [inline lib/docx/body-reader.js]: I think this breaks the behaviour of runs since it replaces the run with a hyperlink, rather than creating a run that contains a hyperlink. For instance, this would mean that bold formatting would be ignored.

File: test/docx/body-reader.tests.js [hunk 6]
```diff
@@ -141,6 +141,18 @@ test("numbering properties are ignored if w:numId is missing", function() {
...
+    var instrText = readXmlElementValue(instrTextXml);
+    assert.deepEqual(instrText, []);
+});
+
+test("run is wrapped in a hyperlink if the runStyle is Hyperlink", function() {
+    var hyperlink = createHyperlink("http://example.com");
+    assert.equal(hyperlink.type, 'hyperlink');
+    assert.equal(hyperlink.href, 'http://example.com');
+});
+
```
Comment (mwilliamson) [inline test/docx/body-reader.tests.js]: I think there's a test case missing for ensuring that elements after the end don't get turned into hyperlinks.

**Comment by jdlehman** [comment]
Thanks for all the great feedback! I definitely agree that the global state is undesirable, it works and is the simplest solution, so I wanted to get that up and get some feedback before I made "deeper" changes.

Before I go into what I am thinking, it is good to note that in addition to the `w:fldChar`s specifying `begin`/`end` of complex fields, a complex field can contain one or many `w:instrText` of data that pertain to that complex field. Hyperlinks only require a single `w:instrText`, but I figure we make this functionality flexible to enable later features pertaining to complex fields to be easily implemented if needed.

Essentially I think we need to create a new type, a `ComplexFieldScope` (or something like that), which is just a collection of data contained in multiple runs (including `w:fldChar` and `w:instrText`). Conceptually this would let us scope state to where it belongs without relying on a global. Look at the following simplified XML structure to see the general approach I was thinking about.

```xml
<Paragraph>
  <ComplexFieldScope data=[" HYPERLINK "example.org"]>
    <run styleName="Hyperlink" styleId="Hyperlink">
      <Text>this is a hyperlink for example.org</Text>
    </run>
  </ComplexFieldScope>
  <!-- more stuff outside of the complex field scope -->
</Paragraph>
```

The above approach could work well, except I noticed that complex fields are not necessarily contained to a single XML tree, eg. they can run across paragraph boundaries if the hyperlink is multiple lines or even sometimes when it appears at the end of the line. So essentially the following can happen, (not valid XML, but conceptually where the scope's bounds are)

```xml
<Paragraph>
  <ComplexFieldScope>
</Paragraph>
<Paragraph>
  </ComplexFieldScope>
</Paragraph>
```

We cannot wrap, both paragraphs in the complex field scope because they might also contain other complex fields.. For that reason I think there will need to be a way to pass the existing complex field scope to the next tree if it has not yet ended. So if we had a multiline hyperlink or any complex field that went across trees it might look like:

```xml
<Paragraph>
  <ComplexFieldScope data=[" HYPERLINK "example.org"]>
    <run styleName="Hyperlink" styleId="Hyperlink">
      <Text>this is the first line of</Text>
    </run>
  </ComplexFieldScope>
</Paragraph>
<Paragraph>
  <ComplexFieldScope data=[" HYPERLINK "example.org"]>
    <run styleName="Hyperlink" styleId="Hyperlink">
      <Text>a multiline hyperlink</Text>
    </run>
  </ComplexFieldScope>
  <!-- more stuff outside of the complex field scope -->
</Paragraph>
```

So there are a few basic approaches I can think of:

1. global state like it is doing now. this is the simplest, but as mentioned global state is not ideal.
2.  existing data structures not only contain children elements and their own data/properties, but might also contain complex field data if they lie within the `begin`/`end` blocks. This breaks down a bit since the complex field data might only be scoped to a subset of the elements children so a similar technique of clearing the current scoped data when an `end` block is reached might be needed.

File: lib/docx/body-reader.js (hunk 2 above)
Comment (jdlehman) [inline lib/docx/body-reader.js]: is `styleId` better to use for this sort of logic? Essentially the only way to tell if the text contained in a run is a hyperlink (from what I can tell based on the spec and from playing around in Word) is to have a `w:rStyle` tag with the value of `Hyperlink`.

Comment (mwilliamson) [inline lib/docx/body-reader.js]: Can you not use the fact that the text of `w:instrText` starts with `HYPERLINK`? Relying on the style at all feels brittle to me.

**Comment by mwilliamson** [comment]
I think there's a third option: the introduction of some local state that tracks whether we're in a hyperlink. This is roughly the same as storing the global state, except that the state is scoped to each read.

This could be threaded through as a second argument to each of the functions that does the reading, but this might be somewhat tedious. An alternative would be to put the state as a local and have all of the readers close over that. In other words, changing `BodyReader` to have some local state (and also renaming it to something else and wrapping it in a stateless `BodyReader` so that it's still not stateful). Something like:

```javascript
function BodyReader(options) {
    return {
         readXmlElement: function readXmlElement(element) {
              return new BodyRead(options).readXmlElement(element);
         },
         ...
    }
}

function BodyRead(options) {
    var state = {};
    // Mostly as the body of BodyReader is now
}
```
This means that the state is scoped to each read, so it still appears stateless to the caller.

I'm not sure which I prefer -- I normally prefer explicit state (i.e. the first option), but it can be a little tedious to write in some cases.

**Comment by mwilliamson** [comment]
Oh, and the naming needs thinking about too!

**Comment by jdlehman** [comment]
I pushed up some refactoring based on some of your earlier comments and moved the state from being global to existing inside `BodyReader`. The other notable change is that `complexFieldData` is an array as the spec indicates that there can be multiple field codes (`w:instrText` nodes). This is to allow for any future work that is done with complex fields.

I am going to take another pass as I think I can clean it up some more and then I will tag you for more feedback.

**Comment by mwilliamson** [comment]
Are you still planning to take another pass, or would you like some feedback?

**Comment by jdlehman** [comment]
It's been a busy couple of days for me. I was planning on taking another pass today and then I would love some feedback 😄 . I'll mention you once I do so you get a notification

**Comment by jdlehman** [comment]
Actually scratch that, I would love some feedback on what is currently up. I was trying to remove the `complexFieldData` state out of `BodyReader`'s scope and pass this data explicitly as an argument. This led down a rabbit hole of needing to add a `complexFieldData` argument to almost every function in `body-reader.js`...

The real issue is that a singe complex field can be defined/used across different XML trees, which I believe Microsoft did such that they can have shared state across trees and partial trees.

There are probably several ways to actually implement the above, but from what I can tell, there would be several repercussions to achieve the same affect as scoping the data to `BodyReader` such that `complexFieldData` is accessible by those that might need it:
- `readXmlElements`, `readXmlElement`, and all `xmlElementReaders` handlers would need to accept this state as a second argument
- rather than returning an element (or in `readXmlElements` case, an array of elements), the complex field state will also need to be returned since complex fields can be defined across the bounds of the XML trees they began in and other children/siblings might need this state

These repercussions would add a lot of complexity to handle the couple of complex fields we might want to support (right now just the other hyperlink format), so I definitely lean towards the simpler approach as it keeps the body reader code readable and easier to understand. That said, if you still feel that having the explicit arguments is better than shared state in the `BodyReader` function scope, feel free to make changes on top of this PR. For other comments/suggestions, I am more than happy to make changes.

As another note, I have been using these changes on another branch in a production application, and have been able to handle both types of hyperlinks.

**Comment by mwilliamson** [comment]
I've pushed a few commits, which should cover some of the previous comments I've made (which seems easier than describing in prose). In particular, the changes:

* hide the statefulness of the body reader from its callers by wrapping it in a stateless shim
* remove the magic indices from the tests
* split up some of the test cases to make it clearer that there are separate behaviours in each
* don't rely on styles to parse hyperlinks

What do you think?

I also went through the spec, and spotted a few things. These feel mostly like edge cases, so feel free to take a look at them, otherwise I'm happy to put them on top of your changes.

* Complex fields can be nested, which isn't currently handled. In particular, if a hyperlink is an outer field, then it will end when any child field ends.

* It looks like `<w:instrText>` elements can be split e.g.

  ```xml
  <w:r>
    <w:instrText>H</w:instrText>
  </w:r>
  <w:r>
    <w:instrText>YPERLINK "https://example.com"</w:instrText>
  </w:r>
  ```

  I've not seen it for hyperlinks in the wild, so I'm not massively concerned, but it should be relatively easy to implement. I think this'll also require parsing `<w:fldChar w:fldCharType="separate"/>`.

* How is the URI encoded? For instance, what happens if the URI contains double quotes?

**Comment by jdlehman** [comment]
Your changes look good to me overall. Thanks for breaking them up into distinct commits, made it easy to see what was changed. I especially like your test changes, they are MUCH more readable.

I will take a look at handling these edge cases tonight. The only question I had is about your change to remove the generalization of complex field data and narrow it to only hyperlink complex field data. Was there a reason you made this more specific. I envisioned storing all complex field data to make it easy for any future work to hook into this data, but maybe that is a bit premature.

**Comment by mwilliamson** [comment]
Just trying to stick to the philosophy of "the simplest thing that works". It makes it clear only that one case is handled, and I tend to find generalisations built around one case tend to not work when trying to make them fit a second or third case (unless it's a problem you've solved before).

**Comment by jdlehman** [comment]
That makes total sense. I think keeping it simple is probably the right call.

I opened another PR against this one to handle these edge cases. I wrote up more details there. https://github.com/jdlehman/mammoth.js/pull/1

**Comment by mwilliamson** [comment]
Thanks both for the initial pull request and responding to the feedback, it's much appreciated!

**Comment by mwilliamson** [comment]
In case it's of interest, I think I managed to clean up the tests a bit by extracting some matchers: you can see the result on master. I also found a bug in the handled of nested fields, which should be fixed now as well.

**Comment by jdlehman** [comment]
Great find on the bug, and thanks for all the feedback along the way and such a great library!
